### PR TITLE
Implement C++ header code generation

### PR DESCRIPTION
**Pull request by shadaj**
TODO:
- [x] Wrap writable-out APIs with ones returning strings instead
- [x] Write example code using the generated APIs
- [x] Complete support for wrapping non-opaque structs
- [x] Topologically sort non-opaque structs
- [x] Emit optional types for results properly
- [x] Take `std::string` as input instead of `char*` with length

### CODE REVIEW

File: tool/src/cpp/mod.rs [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+            }
+            gen_type(underlying.as_ref(), true, env, out)?;
+            write!(out, "&")?;
+            if behind_ref {
+                write!(out, "*")?;
+            }
+        }
+
+        ast::TypeName::Primitive(prim) => {
+            let prim_type = match prim {
```
Comment (shadaj) [inline tool/src/cpp/mod.rs:208]: TODO(shadaj): move this into shared logic

File: example/cpp/api.hpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
+#include <stdint.h>
+#include <stddef.h>
+#include <stdbool.h>
+#include <memory>
+
+namespace capi {
```
Comment (Manishearth) [inline example/cpp/api.hpp:6]: followup: worth filing an issue for making all of this customizeable

File: example/cpp/api.hpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+struct ICU4XDataProviderDeleter {
+  void operator()(capi::ICU4XDataProvider* l) const noexcept {
+    capi::ICU4XDataProvider_destroy(l);
+  }
+};
+class ICU4XDataProvider {
+ public:
+  ICU4XDataProvider ICU4XDataProvider_new_static() {
+    return ICU4XDataProvider(capi::ICU4XDataProvider_new_static());}
```
Comment (Manishearth) [inline example/cpp/api.hpp:19]: nit: formatting

Comment (Manishearth) [reply example/cpp/api.hpp:19]: here and elsewhere

File: example/cpp/api.hpp (hunk 2 above)
Comment (shadaj) [reply example/cpp/api.hpp:8]: #27

File: .vscode/settings.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (Manishearth) [inline .vscode/settings.json:1]: probably don't check this in :smile:

File: example/cpp/api.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+typedef struct ICU4XFixedDecimalFormatResult {
+    ICU4XFixedDecimalFormat* fdf;
+    bool success;
+} ICU4XFixedDecimalFormatResult;
+
+ICU4XDataProvider* ICU4XDataProvider_new_static();
+void ICU4XDataProvider_destroy(ICU4XDataProvider* self);
+
+ICU4XFixedDecimal* ICU4XFixedDecimal_new(int32_t v);
```
Comment (sffc) [inline example/cpp/api.h:31]: `create` not `new` right?

**Review by Manishearth — APPROVED** [review]

**Pull request merged**